File: Core/src/main/java/com/hqq/core/ui/web/BaseWebFragment.java
package com.hqq.core.ui.web;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.Intent;
import android.content.res.ColorStateList;
import android.graphics.Bitmap;
import android.os.Build;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.View;
import android.webkit.URLUtil;
import android.webkit.WebChromeClient;
import android.webkit.WebResourceError;
import android.webkit.WebResourceRequest;
import android.webkit.WebResourceResponse;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.webkit.WebViewClient;
import android.widget.ProgressBar;

import androidx.annotation.Nullable;

import com.hqq.core.R;
import com.hqq.core.listenner.ScriptInterface;
import com.hqq.core.listenner.WebLoadListener;
import com.hqq.core.ui.BaseFragment;
import com.hqq.core.utils.BundleUtils;
import com.hqq.core.utils.RegexUtils;
import com.hqq.core.utils.VersionUtils;

/**
 * @Author : huangqiqiang
 * @Package : com.qi.core.ui
 * @FileName :   BaseWebFragment
 * @Date : 2019/3/4 0004  下午 4:22
 * @Email : qiqiang213@gmail.com
 * @Descrive :
 */
public class BaseWebFragment extends BaseFragment {

    private ProgressBar mProgressBar;
    private WebView mWebView;
    private String mUrl;
    private CharSequence mTitle = "";
    private WebLoadListener mWebLoadListener;
    private ColorStateList mProgressBarColor;
    private ScriptInterface mScriptInterface;

    public static BaseWebFragment instantiate(Context context, String title, String url) {
        return instantiate(context, title, url, null);
    }

    public static BaseWebFragment instantiate(Context context, String title, String url, ScriptInterface scriptInterface) {
        BaseWebFragment baseWebFragment = new BaseWebFragment();
        Bundle bundle = new Bundle();
        bundle.putString(context.getString(R.string.key_url), url);
        bundle.putString(context.getString(R.string.key_title), title);
        baseWebFragment.setArguments(bundle);
        baseWebFragment.setScriptInterface(scriptInterface);
        return baseWebFragment;
    }

    @Override
    public int getViewId() {
        return R.layout.fragment_web;
    }

    @Override
    public void onPause() {
        super.onPause();
        mWebView.onPause();
    }

    @Override
    public void onResume() {
        super.onResume();
        mWebView.onResume();
    }

    @Override
    public void initDefConfig() {
        super.initDefConfig();
        mRootViewBuild.setShowStatusBar(true);
        mRootViewBuild.setShowToolBar(true);
    }

    @SuppressLint("JavascriptInterface")
    @Override
    public void initView() {
        mWebView = (WebView) findViewById(R.id.web_view);
        mProgressBar = (ProgressBar) findViewById(R.id.pb_progressbar);
        if (RegexUtils.unNull(mProgressBarColor)) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                mProgressBar.setIndeterminateTintList(mProgressBarColor);
            }

        }

        mProgressBar.setMax(100);
        //先加载5%，以使用户觉得界面没有卡死
        mProgressBar.setProgress(5);


        initWebViewSettings();

        mWebView.setWebViewClient(getWebViewClient());
        mWebView.setWebChromeClient(getWebChromeClient());
        // 标识 为Android的 js 支持 对象是activity
        if (mScriptInterface != null) {
            mWebView.addJavascriptInterface(getActivity(), "android");
        }

        mUrl = BundleUtils.getString(this, getString(R.string.key_url));
        mTitle = BundleUtils.getString(this, getString(R.string.key_title));
        getActivity().setTitle(mTitle);
        mWebView.loadUrl(mUrl);

    }

    protected void initWebViewSettings() {
        WebSettings settings = mWebView.getSettings();
        settings.setUserAgentString(settings.getUserAgentString() + "" + mActivity.getPackageName());
        settings.setJavaScriptEnabled(true);
        settings.setJavaScriptCanOpenWindowsAutomatically(true);
        settings.setAllowFileAccess(true);
        settings.setUseWideViewPort(true);
        settings.setLoadWithOverviewMode(true);
        settings.setBuiltInZoomControls(true);
        settings.setDomStorageEnabled(true);

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.HONEYCOMB) {
            settings.setDisplayZoomControls(false);
        }
        // MIXED_CONTENT_ALWAYS_ALLOW：允许从任何来源加载内容，即使起源是不安全的；
        // MIXED_CONTENT_NEVER_ALLOW：不允许Https加载Http的内容，即不允许从安全的起源去加载一个不安全的资源；
        // MIXED_CONTENT_COMPATIBILITY_MODE：当涉及到混合式内容时，WebView 会尝试去兼容最新Web浏览器的风格。
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            mWebView.getSettings().setMixedContentMode(WebSettings.MIXED_CONTENT_ALWAYS_ALLOW);
        }

        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.JELLY_BEAN_MR2) {
            //提高渲染优先级
            settings.setRenderPriority(WebSettings.RenderPriority.HIGH);
            settings.setSavePassword(false);
            settings.setPluginState(WebSettings.PluginState.ON);
        }
    }

    protected WebViewClient getWebViewClient() {
        return new WebViewClient() {
            @Override
            public boolean shouldOverrideUrlLoading(WebView view, WebResourceRequest request) {
                //处理 android err_unknown_url_scheme异常
                if (URLUtil.isNetworkUrl(request.getUrl().toString())) {
                    return false;
                }
                if (VersionUtils.appInstalledOrNot(getActivity(), request.getUrl().toString())) {
                    Intent intent = new Intent(Intent.ACTION_VIEW, request.getUrl());
                    startActivity(intent);
                } else {
                    // do something if app is not installed
                }


                return super.shouldOverrideUrlLoading(view, request);

            }

            @Nullable
            @Override
            public WebResourceResponse shouldInterceptRequest(WebView view, WebResourceRequest request) {
                return super.shouldInterceptRequest(view, request);
                //request.getUrl()
            }

            @Override
            public void onPageStarted(WebView view, String url, Bitmap favicon) {
                super.onPageStarted(view, url, favicon);
                mProgressBar.setVisibility(View.VISIBLE);
                if (null != mWebLoadListener) {
                    mWebLoadListener.onPageStarted(url, favicon);
                }
            }

            @Override
            public void onPageFinished(WebView view, String url) {
                super.onPageFinished(view, url);
                mProgressBar.setVisibility(View.GONE);
                if (null != mWebLoadListener) {
                    mWebLoadListener.onPageFinished(url);
                }
            }

            @Override
            public void onReceivedError(WebView view, WebResourceRequest request, WebResourceError error) {
                super.onReceivedError(view, request, error);
                mProgressBar.setVisibility(View.GONE);

            }

        };
    }

    protected WebChromeClient getWebChromeClient() {
        return new WebChromeClient() {

            @Override
            public void onReceivedTitle(WebView view, String title) {
                super.onReceivedTitle(view, title);
                if (TextUtils.isEmpty(mTitle) && RegexUtils.unNull(title) && mRootViewBuild != null && mRootViewBuild.getDefToolBar() != null) {
                    if (null != mRootViewBuild && null != mRootViewBuild.getDefToolBar()) {
                        mRootViewBuild.getDefToolBar().setToolbarTitle(title);
                    }
                }
            }

            @Override
            public void onProgressChanged(WebView view, int newProgress) {
                if (newProgress >= 5) {
                    mProgressBar.setProgress(newProgress);
                }
                super.onProgressChanged(view, newProgress);
            }
        };
    }

    public final boolean onBackPressed() {
        if (mWebView.canGoBack()) {
            mWebView.goBack();
            return true;
        }
        return false;
    }

    public final void loadUrl(String url) {
        loadUrl(url, true);
    }

    public final void loadUrl(String url, boolean clear) {
        if (mWebView != null) {
            if (clear) {
                mWebView.clearHistory();
            }
            mWebView.loadUrl(mUrl = url);
        }
    }

    public BaseWebFragment setWebLoadListener(WebLoadListener webLoadListener) {
        mWebLoadListener = webLoadListener;
        return this;
    }

    public void setScriptInterface(ScriptInterface scriptInterface) {
        mScriptInterface = scriptInterface;
    }

    public ProgressBar getProgressBar() {
        return mProgressBar;
    }

    public final WebView getWebView() {
        return mWebView;
    }

    public void setProgressBar(ProgressBar progressBar) {
        mProgressBar = progressBar;
    }

    public ColorStateList getProgressBarColor() {
        return mProgressBarColor;
    }

    public void setProgressBarColor(ColorStateList progressBarColor) {
        mProgressBarColor = progressBarColor;
    }


}

File: Core/src/main/java/com/hqq/core/ui/web/BaseWebFragment.kt
package com.hqq.core.ui.web

import android.annotation.SuppressLint
import android.content.Context
import android.content.Intent
import android.content.res.ColorStateList
import android.graphics.Bitmap
import android.os.Build
import android.os.Bundle
import android.view.View
import android.webkit.*
import android.widget.ProgressBar
import com.hqq.core.R
import com.hqq.core.listenner.ScriptInterface
import com.hqq.core.listenner.WebLoadListener
import com.hqq.core.ui.base.BaseFragment
import com.hqq.core.utils.BundleUtils
import com.hqq.core.utils.RegexUtils
import com.hqq.core.utils.VersionUtils

/**
 * @Author : huangqiqiang
 * @Package : com.qi.core.ui
 * @FileName :   BaseWebFragment
 * @Date : 2019/3/4 0004  下午 4:22
 * @Email : qiqiang213@gmail.com
 * @Descrive :
 */
class BaseWebFragment : BaseFragment() {
    var progressBar: ProgressBar? = null
    var webView: WebView? = null
        private set
    private var mUrl: String? = null
    private var mTitle: CharSequence? = ""
    private var mWebLoadListener: WebLoadListener? = null
    var progressBarColor: ColorStateList? = null
    private var mScriptInterface: ScriptInterface? = null

    override val layoutViewId: Int = R.layout.fragment_web


    override fun onPause() {
        super.onPause()
        webView!!.onPause()
    }

    override fun onResume() {
        super.onResume()
        webView!!.onResume()
    }

    override fun initConfig() {
        super.initConfig()
        rootViewBuild.mIRootViewImpl.iToolBarBuilder.showToolBar=true
        rootViewBuild.mIRootViewImpl.iToolBarBuilder.showStatusBar=true
    }

    @SuppressLint("JavascriptInterface")
    override fun initView() {
        webView = findViewById(R.id.web_view) as WebView?
        progressBar = findViewById(R.id.pb_progressbar) as ProgressBar?
        if (RegexUtils.unNull(progressBarColor)) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                progressBar!!.indeterminateTintList = progressBarColor
            }
        }
        progressBar!!.max = 100
        //先加载5%，以使用户觉得界面没有卡死
        progressBar!!.progress = 5
        initWebViewSettings()
        webView!!.webViewClient = webViewClient
        webView!!.webChromeClient = webChromeClient
        // 标识 为Android的 js 支持 对象是activity
        if (mScriptInterface != null) {
            webView!!.addJavascriptInterface(requireActivity(), "android")
        }
        mUrl = BundleUtils.getString(this, getString(R.string.key_url))
        mTitle = BundleUtils.getString(this, getString(R.string.key_title))
        requireActivity().title = mTitle
        iToolBar?.setToolbarTitle(mTitle)
        webView!!.loadUrl(mUrl!!)
    }

    protected fun initWebViewSettings() {
        val settings = webView!!.settings
        settings.setUserAgentString(settings.userAgentString + "" + requireActivity().packageName)
        settings.javaScriptEnabled = true
        settings.javaScriptCanOpenWindowsAutomatically = true
        settings.allowFileAccess = true
        settings.useWideViewPort = true
        settings.loadWithOverviewMode = true
        settings.builtInZoomControls = true
        settings.domStorageEnabled = true
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.HONEYCOMB) {
            settings.displayZoomControls = false
        }
        // MIXED_CONTENT_ALWAYS_ALLOW：允许从任何来源加载内容，即使起源是不安全的；
        // MIXED_CONTENT_NEVER_ALLOW：不允许Https加载Http的内容，即不允许从安全的起源去加载一个不安全的资源；
        // MIXED_CONTENT_COMPATIBILITY_MODE：当涉及到混合式内容时，WebView 会尝试去兼容最新Web浏览器的风格。
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            webView!!.settings.mixedContentMode = WebSettings.MIXED_CONTENT_ALWAYS_ALLOW
        }
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.JELLY_BEAN_MR2) {
            //提高渲染优先级
            settings.setRenderPriority(WebSettings.RenderPriority.HIGH)
            settings.savePassword = false
            settings.pluginState = WebSettings.PluginState.ON
        }
    }//request.getUrl()// do something if app is not installed

    //处理 android err_unknown_url_scheme异常
    protected val webViewClient: WebViewClient
        protected get() = object : WebViewClient() {
            override fun shouldOverrideUrlLoading(view: WebView, request: WebResourceRequest): Boolean {
                //处理 android err_unknown_url_scheme异常
                if (URLUtil.isNetworkUrl(request.url.toString())) {
                    return false
                }
                if (VersionUtils.appInstalledOrNot(activity, request.url.toString())) {
                    val intent = Intent(Intent.ACTION_VIEW, request.url)
                    startActivity(intent)
                } else {
                    // do something if app is not installed
                }
                return super.shouldOverrideUrlLoading(view, request)
            }

            override fun shouldInterceptRequest(view: WebView, request: WebResourceRequest): WebResourceResponse? {
                return super.shouldInterceptRequest(view, request)
                //request.getUrl()
            }

            override fun onPageStarted(view: WebView, url: String, favicon: Bitmap) {
                super.onPageStarted(view, url, favicon)
                progressBar!!.visibility = View.VISIBLE
                if (null != mWebLoadListener) {
                    mWebLoadListener!!.onPageStarted(url, favicon)
                }
            }

            override fun onPageFinished(view: WebView, url: String) {
                super.onPageFinished(view, url)
                progressBar!!.visibility = View.GONE
                if (null != mWebLoadListener) {
                    mWebLoadListener!!.onPageFinished(url)
                }
            }

            override fun onReceivedError(view: WebView, request: WebResourceRequest, error: WebResourceError) {
                super.onReceivedError(view, request, error)
                progressBar!!.visibility = View.GONE
            }
        }

    protected val webChromeClient: WebChromeClient
        protected get() = object : WebChromeClient() {
            override fun onReceivedTitle(view: WebView, title: String?) {
                super.onReceivedTitle(view, title)
                title?.let {
                    if (mTitle.isNullOrEmpty()) {
                        iToolBar?.setToolbarTitle(mTitle)
                    }
                }
            }

            override fun onProgressChanged(view: WebView, newProgress: Int) {
                if (newProgress >= 5) {
                    progressBar!!.progress = newProgress
                }
                super.onProgressChanged(view, newProgress)
            }
        }

    fun onBackPressed(): Boolean {
        if (webView!!.canGoBack()) {
            webView!!.goBack()
            return true
        }
        return false
    }

    @JvmOverloads
    fun loadUrl(url: String, clear: Boolean = true) {
        if (webView != null) {
            if (clear) {
                webView!!.clearHistory()
            }
            webView!!.loadUrl(url.also { mUrl = it })
        }
    }

    fun setWebLoadListener(webLoadListener: WebLoadListener?): BaseWebFragment {
        mWebLoadListener = webLoadListener
        return this
    }

    fun setScriptInterface(scriptInterface: ScriptInterface?) {
        mScriptInterface = scriptInterface
    }

    companion object {
        @JvmOverloads
        fun instantiate(context: Context, title: String?, url: String?, scriptInterface: ScriptInterface? = null): BaseWebFragment {
            val baseWebFragment = BaseWebFragment()
            val bundle = Bundle()
            bundle.putString(context.getString(R.string.key_url), url)
            bundle.putString(context.getString(R.string.key_title), title)
            baseWebFragment.arguments = bundle
            baseWebFragment.setScriptInterface(scriptInterface)
            return baseWebFragment
        }
    }
}